Logging & Testing

tmck-code opened the pull request:
## Context

During my adventures in this repository, I've found myself often adding a hacky log line in order to gain insight into what's going on.

- A log file is invaluable when developing, especially when there are few/no tests
- A log file can help debug issues for users if they submit it

As I was implementing the log module, I found myself wishing that I could write some tests instead of doing trial & error via running the main binary.

- [Context](#context)
- [Changes](#changes)
  - [Logger](#logger)
    - [Using in code](#using-in-code)
    - [Enabling via config](#enabling-via-config)
    - [Notes](#notes)
  - [Testing](#testing)

---

## Changes

### Logger

#### Using in code

I added a logger under `durdraw/log.py`. I should hopefully be fairly straightforward to use. If you'd like to add a log in a module that doesn't have it already, you'll need to

- Instantiate a new logger, giving it a `name` that matches the module you're in
- Then log with it!

e.g.

```python
import durdraw.log as log

# instantiate a logger outside any class, e.g.
LOG = log.getLogger('UndoHistory')

# or, instantiate it inside a class init method, e.g.
class UndoHistory:
    def __init__(self):
       ...
       self.log = log.getLogger('UndoHistory')
```

then, you can log with it in a couple of ways:

```python
# with a plain message
log.info('Hello, world!')
# {"timestamp": "2024-12-09T20:17:25.787143+11:00", "level": "INFO", "name": "durdraw.UndoHistory", "msg": "Hello, world!", "data": {}}

# with additional key/value data
log.info('Hello, world!', {'world': 'mars'})
# {"timestamp": "2024-12-09T20:17:25.787143+11:00", "level": "INFO", "name": "durdraw.UndoHistory", "msg": "Hello, world!", "data": {"world": "mars"}}
```

#### Enabling via config

To enable, set the following in the `~/.durdraw/durdraw.ini` config file:

```
[Logging]
level: INFO
; you can change filepath to change the log file
; filepath: ./durdraw.log 
```

This will enable all INFO logs (or higher severity) to be logged to `./durdraw.log`

#### Notes

- No log file will be created if no logs are emitted

Why Structured Logging?

![image](https://github.com/user-attachments/assets/86a2e83b-ac9a-4731-96c4-30f0e3c4e09b)

### Testing

- The tests are placed under `test/`.
- There is a single test module for the log
- Run tests with `pytest -vv test/`
  - Added this instruction to the readme, along with pytest install instructions

#### pytest vs unittest

functionality | pytest | unittest
-- | -- | --
display non-truncated diff | use -vv cli flag when running | add self.maxDiff = None to test class
display n slowest tests | use cli flag --durations=N | can't do it (it can now as of 3.12)
expectation style | assert, roll your own! | Heaps of hyper-specific comparison methods e.g.assertIsInstance() Vs assert is instance()
print name of currently running test | use -v or -vv | can't
See stdout output | use -s | only for failed tests, need to adjust code

Overall, it looks like the unittest library has undergone a _lot_ of improvements recently, closing the gap between it and pytest. If you prefer I can use unittest instead 🤙

cmang commented:
Nice!

> During my adventures in this repository

Me at some point: "Gee, nobody will ever read this sleep-deprived brute force code..."

I noticed that the tests fail and Durdraw won't start with Python 3.9, due to using typing.TypeAlias in log.py.

Looking at the docs, it looks like they added TypeAlias in 3.10 and [deprecated it in 3.12](https://docs.python.org/3/library/typing.html#typing.TypeAlias).

To me that seems like kind of a short window of compatibility to be adding to the code (or to Python itself, for that matter).  Maybe enough that it's worth replacing in this case, or only using TypeAlias if the Python version is correct.  

For example, the Python 3.9.6 that _still_ comes with macOS Sonoma fails.  A version released in June 2021... pretty old, but it's out there. 

I understand we probably shouldn't support 3.9 forever, but I did write some patches earlier this year to keep it working. If Macs didn't ship with 3.9 I probably wouldn't care.  But until Apple updates the version of Python built into macOS, I'd like to keep supporting it. And since ASCII art is so universal, I prefer for Durdraw philosophically to try to scale to systems new and old, big and small. :) 

Besides that... it looks cool! I generated some debug messages, and the output is nice. And I guess the tests are already helping us catch stuff.

Thank you, Tom!

```
 % /usr/bin/python3 start-durdraw
Traceback (most recent call last):
  File "/Users/samfoster/src/durdraw-tmck-code/durdraw/start-durdraw", line 5, in <module>
    from durdraw import main
  File "/Users/samfoster/src/durdraw-tmck-code/durdraw/durdraw/main.py", line 12, in <module>
    from durdraw.durdraw_appstate import AppState
  File "/Users/samfoster/src/durdraw-tmck-code/durdraw/durdraw/durdraw_appstate.py", line 12, in <module>
    import durdraw.durdraw_file as durfile
  File "/Users/samfoster/src/durdraw-tmck-code/durdraw/durdraw/durdraw_file.py", line 11, in <module>
    from durdraw.durdraw_movie import Movie
  File "/Users/samfoster/src/durdraw-tmck-code/durdraw/durdraw/durdraw_movie.py", line 3, in <module>
    import durdraw.log as log
  File "/Users/samfoster/src/durdraw-tmck-code/durdraw/durdraw/log.py", line 30, in <module>
    from typing import TypeAlias, Literal
ImportError: cannot import name 'TypeAlias' from 'typing' (/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/typing.py)
```

```
% pytest test
====================================================================== test session starts ======================================================================
platform darwin -- Python 3.9.12, pytest-7.1.2, pluggy-1.0.0
rootdir: /Users/samfoster/src/durdraw-tmck-code/durdraw
plugins: anyio-3.5.0
collected 0 items / 1 error

============================================================================ ERRORS =============================================================================
___________________________________________________________ ERROR collecting test/durdraw/test_log.py ___________________________________________________________
ImportError while importing test module '/Users/samfoster/src/durdraw-tmck-code/durdraw/test/durdraw/test_log.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../miniconda3/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
test/durdraw/test_log.py:1: in <module>
    import durdraw.log as log
durdraw/log.py:30: in <module>
    from typing import TypeAlias, Literal
E   ImportError: cannot import name 'TypeAlias' from 'typing' (/Users/samfoster/miniconda3/lib/python3.9/typing.py)
==================================================================== short test summary info ====================================================================
ERROR test/durdraw/test_log.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================================= 1 error in 0.05s ========================================================================
```

tmck-code commented:
> I noticed that the tests fail and Durdraw won't start with Python 3.9, due to using typing.TypeAlias in log.py.

Oh gosh that totally slipped past me! The type alias stuff is totally optional, I'll just remove it ✂️

Thanks for the heads up re: python 3.9. According to the [fully sweet version page](https://devguide.python.org/versions/), it'll be supported with security updates until the end of 2025, so not going anywhere for now.

I'll make sure to do any testing and development using python 3.9 to ensure that there aren't any surprises.

Re the testing lib: a few of the upgrades to the unittest library are >3.9, so I'll keep rolling with pytest for now (it's not overly difficult to switch in future)

cmang commented:
> Thanks for the heads up re: python 3.9. According to the [fully sweet version page](https://devguide.python.org/versions/), it'll be supported with security updates until the end of 2025, so not going anywhere for now.

Nice useful page. It looks like I was partially wrong about macOS. MacOS Sequoia (released this September) did ship with an updated Python (3.13), though upgrading the OS from Sonoma (14.x) to Sequoia (15.x) did not upgrade the system Python install for me. Still version 3.9.6. If that's the case for other people, then most Macs are probably not running 3.13 yet.

> I'll make sure to do any testing and development using python 3.9 to ensure that there aren't any surprises.

Looks good. I am now getting a test fail on both 3.9 and 3.12 from test_log_timestamp_timezone.  Durdraw itself seems to run, though.

```
% python -m pytest -vv test
====================================================================== test session starts ======================================================================
platform darwin -- Python 3.12.3, pytest-8.3.4, pluggy-1.5.0 -- /Users/samfoster/.pyenv/versions/3.12.3/bin/python
cachedir: .pytest_cache
rootdir: /Users/samfoster/src/durdraw-tmck-code/durdraw
collected 6 items

test/durdraw/test_log.py::TestLog::test_log_complete_format PASSED                                                                                        [ 16%]
test/durdraw/test_log.py::TestLog::test_log_timestamp_timezone FAILED                                                                                     [ 33%]
test/durdraw/test_log.py::TestLog::test_log_no_args PASSED                                                                                                [ 50%]
test/durdraw/test_log.py::TestLog::test_log_kwargs PASSED                                                                                                 [ 66%]
test/durdraw/test_log.py::TestLog::test_log_to_file PASSED                                                                                                [ 83%]
test/durdraw/test_log.py::TestLog::test_log_no_emit PASSED                                                                                                [100%]

=========================================================================== FAILURES ============================================================================
______________________________________________________________ TestLog.test_log_timestamp_timezone ______________________________________________________________

self = <test.durdraw.test_log.TestLog object at 0x10181a450>

    def test_log_timestamp_timezone(self):
        logger, fake_stream = init_test_logger()
        logger.info('Hello, world!')

        result = datetime.fromisoformat(json.loads(fake_stream.getvalue())['timestamp'])

        assert result.utcoffset() is not None
>       assert result.utcoffset().seconds == time.localtime().tm_gmtoff
E       assert 64800 == -21600
E        +  where 64800 = datetime.timedelta(days=-1, seconds=64800).seconds
E        +    where datetime.timedelta(days=-1, seconds=64800) = <built-in method utcoffset of datetime.datetime object at 0x101826550>()
E        +      where <built-in method utcoffset of datetime.datetime object at 0x101826550> = datetime.datetime(2024, 12, 10, 11, 20, 58, 506077, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=64800))).utcoffset
E        +  and   -21600 = time.struct_time(tm_year=2024, tm_mon=12, tm_mday=10, tm_hour=11, tm_min=20, tm_sec=58, tm_wday=1, tm_yday=345, tm_isdst=0).tm_gmtoff
E        +    where time.struct_time(tm_year=2024, tm_mon=12, tm_mday=10, tm_hour=11, tm_min=20, tm_sec=58, tm_wday=1, tm_yday=345, tm_isdst=0) = <built-in function localtime>()
E        +      where <built-in function localtime> = time.localtime

test/durdraw/test_log.py:46: AssertionError
----------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------
INFO     durdraw.test_log:test_log.py:41 {"timestamp": "2024-12-10T11:20:58.506077-06:00", "level": "INFO", "name": "durdraw.test_log", "msg": "Hello, world!", "data": {}}
==================================================================== short test summary info ====================================================================
FAILED test/durdraw/test_log.py::TestLog::test_log_timestamp_timezone - assert 64800 == -21600
 +  where 64800 = datetime.timedelta(days=-1, seconds=64800).seconds
 +    where datetime.timedelta(days=-1, seconds=64800) = <built-in method utcoffset of datetime.datetime object at 0x101826550>()
 +      where <built-in method utcoffset of datetime.datetime object at 0x101826550> = datetime.datetime(2024, 12, 10, 11, 20, 58, 506077, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=64800))).utcoffset
 +  and   -21600 = time.struct_time(tm_year=2024, tm_mon=12, tm_mday=10, tm_hour=11, tm_min=20, tm_sec=58, tm_wday=1, tm_yday=345, tm_isdst=0).tm_gmtoff
 +    where time.struct_time(tm_year=2024, tm_mon=12, tm_mday=10, tm_hour=11, tm_min=20, tm_sec=58, tm_wday=1, tm_yday=345, tm_isdst=0) = <built-in function localtime>()
 +      where <built-in function localtime> = time.localtime
================================================================== 1 failed, 5 passed in 0.07s ==================================================================
```

For grins, I also tried running it on an old Debian Buster machine with Python 3.7.3. Which is super old now, and officially unsupported, but I do play with Durdraw on it for fun. It didn't work. Not a big deal, though.

```
$ python3 ./start-durdraw
Traceback (most recent call last):
  File "./start-durdraw", line 5, in <module>
    from durdraw import main
  File "/home/indy/src/durdraw-tmck/durdraw/durdraw/main.py", line 12, in <module>
    from durdraw.durdraw_appstate import AppState
  File "/home/indy/src/durdraw-tmck/durdraw/durdraw/durdraw_appstate.py", line 12, in <module>
    import durdraw.durdraw_file as durfile
  File "/home/indy/src/durdraw-tmck/durdraw/durdraw/durdraw_file.py", line 11, in <module>
    from durdraw.durdraw_movie import Movie
  File "/home/indy/src/durdraw-tmck/durdraw/durdraw/durdraw_movie.py", line 3, in <module>
    import durdraw.log as log
  File "/home/indy/src/durdraw-tmck/durdraw/durdraw/log.py", line 89, in <module>
    def _getLogger(name: str, level: int = logging.CRITICAL, handlers: list[logging.Handler] = []) -> logging.Logger:
TypeError: 'type' object is not subscriptable
```

tmck-code commented:
Phew, quite the journey! 😓

Your test failures look to be caused by some timezone shenanigans on your computer, it's the first time that I've seen those 2 techniques produce a different value (i.e. `64800 == -21600`)
                                                                                                                      Noticing this, I realised that - in the case of users supplying logs for debugging - logs would contain their local timezone, and that it would be better to use a normalised UTC as we wouldn't actually care about the user TZ.                                                                            

I added a .ini option `local-tz` (defaults to false) to toggle using your local tz if you're using the logs to develop/you care about interpreting the times easily

---                                                                                                                         

I updated the tests and log code to be hopefully more robust and work better with older python versions

- use strptime over fromisotimestamp
- don't compare log tz to default user tz
- don't use type hints with built-in types e.g. `list[]`

tmck-code commented:
Hopefully it's all good now! 🤞 Tested with 3.7, 3.9 and 3.12 both the unit tests and running with the ini

cmang reviewed (approved):
Awesome. Thanks for putting up with my weird obsession with backwards compatibility. It's good!

<img width="291" alt="Screenshot 2024-12-10 at 10 01 56 PM" src="https://github.com/user-attachments/assets/ef145917-bb2b-4a0d-aec7-0b2734bdd0d0">

> Your test failures look to be caused by some timezone shenanigans on your computer, it's the first time that I've seen those 2 techniques produce a different value (i.e. 64800 == -21600). 

Maybe it's because my time zone (CST/Chicago) is in Daylight Savings Time? I'm not sure.

> I could make a pipeline and several docker images (or just one) that contains several python versions, and the tests could be run across all those versions to ensure that there are no regressions, and this can be done on every commit/push

Sounds super useful. :)
